rust/src/ffi/base64.rs

```diff
@@ -27,6 +28,90 @@ pub enum Base64ReturnCode {
     SC_BASE64_OVERFLOW,
 }
 
+#[repr(C)]
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum Base64Mode {
+    /* If the following strings were to be passed to the decoder with RFC2045 mode,
+     * the results would be as follows. See the unittest B64TestVectorsRFC2045 in
+     * src/util-base64.c
+     *
+     * BASE64("") = ""
+     * BASE64("f") = "Zg=="
+     * BASE64("fo") = "Zm8="
+     * BASE64("foo") = "Zm9v"
+     * BASE64("foob") = "Zm9vYg=="
+     * BASE64("fooba") = "Zm9vYmE="
+     * BASE64("foobar") = "Zm9vYmFy"
+     * BASE64("foobar") = "Zm 9v Ym Fy"   <-- Notice how the spaces are ignored
+     * BASE64("foobar") = "Zm$9vYm.Fy"    # According to RFC 2045, All line breaks or *other
+     * characters* not found in base64 alphabet must be ignored by decoding software
+     * */
+    Base64ModeRFC2045 = 0, /* SPs are allowed during transfer but must be skipped by Decoder */
+    Base64ModeStrict,
+    /* If the following strings were to be passed to the decoder with RFC4648 mode,
+     * the results would be as follows. See the unittest B64TestVectorsRFC4648 in
+     * src/util-base64.c
+     *
+     * BASE64("") = ""
+     * BASE64("f") = "Zg=="
+     * BASE64("fo") = "Zm8="
+     * BASE64("foo") = "Zm9v"
+     * BASE64("foob") = "Zm9vYg=="
+     * BASE64("fooba") = "Zm9vYmE="
+     * BASE64("foobar") = "Zm9vYmFy"
+     * BASE64("f") = "Zm 9v Ym Fy"   <-- Notice how the processing stops once space is encountered
+     * BASE64("f") = "Zm$9vYm.Fy"    <-- Notice how the processing stops once an invalid char is
+     * encountered
+     * */
+    Base64ModeRFC4648, /* reject the encoded data if it contains characters outside the base alphabet */
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn Base64DecodeBufferSize(input_len: u32) -> u32 {
+    return ((input_len * 3) + (input_len % 4)) / 4;
+}
+
+/// Base64 decode a buffer.
+///
+/// This method exposes the Rust base64 decoder to C and should not be called from
+/// Rust code.
+///
+/// It allows decoding in the modes described by ``Base64Mode`` enum.
+#[no_mangle]
+pub unsafe extern "C" fn Base64Decode(
+    input: *const u8, len: usize, max_decoded: u32, mode: Base64Mode, output: *mut u8) -> u32 {
```

Comment:
What is the purpose of `max_decoded` argument ?

Reply:
This is used particularly by `base64.decode` to indicate the maximum number of bytes we want to decode. I suppose we should consider MIN(encoded_len*3/4, max_decoded) as the actual number of bytes that should be decoded. Looking at the code, I'm not doing that so this needs fixing. Thanks a lot! 🙇🏽‍♀️

Reply:
When is not `max_decoded` equal to `Base64DecodeBufferSize(len)` ?

Reply:
(and my point was rather to have a comment explaining this argument ;-) )

Reply:
> When is not max_decoded equal to Base64DecodeBufferSize(len) ?

Turns out it is always much bigger. So, this arg can be safely removed and I don't need to do any modifications.